tests/framework/e2e/downgrade.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2025 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"go.etcd.io/etcd/api/v3/version"
+	"go.etcd.io/etcd/server/v3/etcdserver"
+	"go.etcd.io/etcd/tests/v3/framework/testutils"
+)
+
+func DowngradeEnable(t *testing.T, epc *EtcdProcessCluster, ver *semver.Version) {
+	t.Logf("etcdctl downgrade enable %s", ver.String())
+	c := epc.Etcdctl()
+	testutils.ExecuteWithTimeout(t, 20*time.Second, func() {
+		err := c.DowngradeEnable(context.TODO(), ver.String())
+		require.NoError(t, err)
+	})
+
+	t.Log("Downgrade enabled, validating if cluster is ready for downgrade")
+	for i := 0; i < len(epc.Procs); i++ {
+		ValidateVersion(t, epc.Cfg, epc.Procs[i], version.Versions{
+			Cluster: ver.String(),
+			Server:  offsetMinor(ver, 1).String(),
+			Storage: ver.String(),
+		})
+		AssertProcessLogs(t, epc.Procs[i], "The server is ready to downgrade")
+	}
+
+	t.Log("Cluster is ready for downgrade")
+}
+
+func DowngradeUpgradeMembers(t *testing.T, lg *zap.Logger, clus *EtcdProcessCluster, numberOfMembersToChange int, currentVersion, targetVersion *semver.Version) error {
+	if lg == nil {
+		lg = clus.lg
+	}
```

Comment:
I think it's better to leave caller to decide logger. This defaulting obscures what logger is used from caller.